scheds/rust/scx_rustlite/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,404 @@
+/* Copyright (c) Andrea Righi <andrea.righi@canonical.com> */
+/*
+ * scx_rustlite: simple user-space scheduler written in Rust
+ *
+ * The main goal of this scheduler is be an "easy to read" template that can be
+ * used to quickly test more complex scheduling policies. For this reason this
+ * scheduler is mostly focused on simplicity and code readability.
+ *
+ * The scheduler is made of a BPF component (dispatcher) that implements the
+ * low level sched-ext functionalities and a user-space counterpart
+ * (scheduler), written in Rust, that implements the actual scheduling policy.
+ *
+ * The BPF dispatcher collects total cputime and weight from the tasks that
+ * need to run, then it sends all details to the user-space scheduler that
+ * decides the best order of execution of the tasks (based on the collected
+ * metrics).
+ *
+ * The user-space scheduler then returns to the BPF component the list of tasks
+ * to be dispatched in the proper order.
+ *
+ * Messages between the BPF component and the user-space scheduler are passed
+ * using two BPF_MAP_TYPE_QUEUE maps: @queued for the messages sent by the BPF
+ * dispatcher to the user-space scheduler and @dispatched for the messages sent
+ * by the user-space scheduler to the BPF dispatcher.
+ *
+ * The BPF dispatcher is completely agnostic of the particular scheduling
+ * policy implemented in user-space. For this reason developers that are
+ * willing to use this scheduler to experiment scheduling policies should be
+ * able to simply modify the Rust component, without having to deal with any
+ * internal kernel / BPF details.
+ *
+ * This software may be used and distributed according to the terms of the
+ * GNU General Public License version 2.
+ */
+#include <scx/common.bpf.h>
+#include "intf.h"
+
+char _license[] SEC("license") = "GPL";
+
+/*
+ * Exit info (passed to the user-space counterpart).
+ */
+int exit_kind = SCX_EXIT_NONE;
+char exit_msg[SCX_EXIT_MSG_LEN];
+
+/*
+ * Scheduler attributes and statistics.
+ */
+u32 usersched_pid; /* User-space scheduler PID */
+static bool usersched_needed; /* Used to wake-up the user-space scheduler */
+const volatile bool switch_partial; /* Switch all tasks or SCHED_EXT tasks */
+const volatile u64 slice_ns = SCX_SLICE_DFL; /* Base time slice duration */
+
+/* Statistics */
+u64 nr_enqueues, nr_user_dispatches, nr_kernel_dispatches;
+u64 nr_sched_congested, nr_tasks_running;
+
+ /* Report additional debugging information */
+const volatile bool debug;
+
+/* Allow to use bpf_printk() only when @debug is set */
+#define dbg_msg(_fmt, ...) do {				\
+	if (unlikely(debug))				\
+		bpf_printk(_fmt, ##__VA_ARGS__);	\
```

Comment:
I _think_ this will fail to compile if there are no args in the debug message? What about this?

`bpf_printk(_fmt __VA_OPT__(,) __VA_ARGS__);`

Reply:
```
#define dbg_msg(_fmt, ...) do {                         \
        if (debug)                                      \
                bpf_printk(_fmt, ##__VA_ARGS__);        \
} while(0)
```
This seems to work also with no args and it doesn't look too ugly. I'll go with this one.

Reply:
Ahh I understand the macro now, it LGTM. Somehow I'd never seen ##__VA_ARGS__ applied that way, but now I'm seeing that it's actually super common.